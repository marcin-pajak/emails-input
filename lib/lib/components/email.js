"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeEmailHandlers = exports.emailTemplate = exports.EMAIL_CLASSES = void 0;
var sanitize_1 = require("../../utils/sanitize");
var typeGuards_1 = require("../../utils/typeGuards");
var constants_1 = require("../constants");
exports.EMAIL_CLASSES = {
    removeButton: 'Email-remove',
    root: 'Email',
    text: 'Email-text',
};
var removeButton = exports.EMAIL_CLASSES.removeButton, root = exports.EMAIL_CLASSES.root, text = exports.EMAIL_CLASSES.text;
/**
 * Email Component Template
 * @param value
 * @param id
 */
exports.emailTemplate = function (value, id) { return "\n<div class=\"" + constants_1.CLASSES.EMAIL + " " + root + " Email--" + (value.isValid ? 'valid' : 'invalid') + "\" data-id=\"" + id + "\">\n  <span class=\"" + text + "\">\n    " + sanitize_1.sanitize(value.email) + "\n  </span>\n  <button aria-label=\"remove email\" class=\"" + removeButton + "\" data-id=\"" + id + "\" type=\"button\">\n    &times;\n  </button>\n</div>"; };
/**
 * Email Component Event Handlers
 * @param handlers
 */
exports.makeEmailHandlers = function (handlers) {
    var removeByLocalId = handlers.removeByLocalId;
    var handleRemoveClick = function (event) {
        var target = event.target;
        if (!typeGuards_1.isClickedButton(target)) {
            return;
        }
        var id = target.dataset.id;
        if (typeof id !== 'undefined') {
            removeByLocalId(id);
        }
        target.blur();
    };
    return {
        handleRemoveClick: handleRemoveClick,
    };
};
//# sourceMappingURL=email.js.map