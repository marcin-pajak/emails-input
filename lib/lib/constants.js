"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLASSES = exports.ChangeType = void 0;
var ChangeType;
(function (ChangeType) {
    ChangeType["Added"] = "Added";
    ChangeType["Removed"] = "Removed";
})(ChangeType = exports.ChangeType || (exports.ChangeType = {}));
exports.CLASSES = {
    CONTAINER: 'EmailsInputWrapper',
    EMAIL: 'EmailsInput-email',
    INPUT: 'EmailsInput-input',
    WRAPPER: 'EmailsInput',
};
//# sourceMappingURL=constants.js.map