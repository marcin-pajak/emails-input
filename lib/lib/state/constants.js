"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LibraryState = exports.ActionType = void 0;
var ActionType;
(function (ActionType) {
    ActionType["Add"] = "ADD";
    ActionType["Remove"] = "REMOVE";
    ActionType["Initialize"] = "INITIALIZE";
    ActionType["Destroy"] = "DESTROY";
    ActionType["UpdateState"] = "UPDATE_STATE";
})(ActionType = exports.ActionType || (exports.ActionType = {}));
var LibraryState;
(function (LibraryState) {
    LibraryState["Initializing"] = "Initializing";
    LibraryState["Initialized"] = "Initialized";
    LibraryState["Destroyed"] = "Destroyed";
})(LibraryState = exports.LibraryState || (exports.LibraryState = {}));
//# sourceMappingURL=constants.js.map