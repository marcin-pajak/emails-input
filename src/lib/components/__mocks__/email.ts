export const emailTemplate = (): string => '<span />';
