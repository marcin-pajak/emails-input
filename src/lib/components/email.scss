.Email {
  box-sizing: border-box;
  color: #050038;
  display: flex;
  font-size: 14px;
  min-height: 24px;
  line-height: 24px;
  margin: 4px;
  min-width: 0;
  user-select: none;
}

.Email--valid {
  border-radius: 100px;
  background: rgba(102, 153, 255, 0.2);
  border: 1px solid rgba(102, 153, 255, 0.2);
  padding-left: 10px;
}

.Email--invalid {
  border: #fff;
  border-bottom: 1px dashed #d92929;
}

.Email-remove {
  background: transparent;
  border: 0;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-size: 17px;
  line-height: 21px;
  margin: 2px;
  padding: 0;
  text-align: center;
  text-decoration: none;
  user-select: none;
  width: 20px;
  vertical-align: middle;
}

.Email-remove:focus,
.Email-remove:hover {
  opacity: 0.8;
  text-decoration: none;
}

.Email-remove:active {
  transform: translate(1px, 1px);
}
