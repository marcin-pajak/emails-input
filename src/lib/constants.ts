export enum ChangeType {
  Added = 'Added',
  Removed = 'Removed',
}

export const CLASSES = {
  CONTAINER: 'EmailsInputWrapper',
  EMAIL: 'EmailsInput-email',
  INPUT: 'EmailsInput-input',
  WRAPPER: 'EmailsInput',
};
