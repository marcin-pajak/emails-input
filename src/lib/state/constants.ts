export enum ActionType {
  Add = 'ADD',
  Remove = 'REMOVE',
  Initialize = 'INITIALIZE',
  Destroy = 'DESTROY',
  UpdateState = 'UPDATE_STATE',
}

export enum LibraryState {
  Initializing = 'Initializing',
  Initialized = 'Initialized',
  Destroyed = 'Destroyed',
}
