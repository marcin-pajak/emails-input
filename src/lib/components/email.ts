import { sanitize } from '../../utils/sanitize';
import { isClickedButton } from '../../utils/typeGuards';
import { EmailActions } from '../actions/types';
import { CLASSES } from '../constants';
import { Entry } from '../state/types';
import { EmailHandlers } from './types';

export const EMAIL_CLASSES = {
  removeButton: 'Email-remove',
  root: 'Email',
  text: 'Email-text',
};

const { removeButton, root, text } = EMAIL_CLASSES;

/**
 * Email Component Template
 * @param value
 * @param id
 */
export const emailTemplate = (value: Entry, id: string): string => `
<div class="${CLASSES.EMAIL} ${root} Email--${
  value.isValid ? 'valid' : 'invalid'
}" data-id="${id}">
  <span class="${text}">
    ${sanitize(value.email)}
  </span>
  <button aria-label="remove email" class="${removeButton}" data-id="${id}" type="button">
    &times;
  </button>
</div>`;

/**
 * Email Component Event Handlers
 * @param handlers
 */
export const makeEmailHandlers = (handlers: EmailActions): EmailHandlers => {
  const { removeByLocalId } = handlers;

  const handleRemoveClick = (event: MouseEvent): void => {
    const { target } = event;

    if (!isClickedButton(target)) {
      return;
    }

    const { id } = target.dataset;

    if (typeof id !== 'undefined') {
      removeByLocalId(id);
    }

    target.blur();
  };

  return {
    handleRemoveClick,
  };
};
