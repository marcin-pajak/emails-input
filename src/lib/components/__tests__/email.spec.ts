import EmailsInput from '../../..';
import { CLASSES } from '../../constants';
import { EMAIL_CLASSES, emailTemplate } from '../email';

describe('email template should', (): void => {
  it('return template for valid email', (): void => {
    expect.hasAssertions();

    expect(
      emailTemplate(
        {
          email: 'valid@email.com',
          isValid: true,
        },
        '1',
      ),
    ).toMatchSnapshot();
  });

  it('return template for invalid email', (): void => {
    expect.hasAssertions();

    expect(
      emailTemplate(
        {
          email: 'invalid',
          isValid: false,
        },
        '1',
      ),
    ).toMatchSnapshot();
  });

  it('return sanitized template', (): void => {
    expect.hasAssertions();

    expect(
      emailTemplate(
        {
          email: '<b>dangerous</b>@email.com',
          isValid: true,
        },
        '1',
      ),
    ).toMatchSnapshot();
  });
});

const ELEMENT_CLASS = 'emails-input';
describe('email component should', (): void => {
  beforeEach((): void => {
    document.body.innerHTML = `<div id="${ELEMENT_CLASS}"></div>`;
  });

  afterEach((): void => {
    const element = document.querySelector(`#${ELEMENT_CLASS}`);
    element?.remove();
  });

  it('handle remove click', (): void => {
    expect.hasAssertions();

    const initialValue = ['one@email.com'];
    const element = document.querySelector(`#${ELEMENT_CLASS}`) as HTMLElement;
    EmailsInput(element, { initialValue });
    expect(document.querySelectorAll(`.${CLASSES.EMAIL}`)).toHaveLength(
      initialValue.length,
    );

    const button = element.querySelector(
      `.${EMAIL_CLASSES.removeButton}`,
    ) as HTMLButtonElement;

    button.click();
    expect(document.querySelectorAll(`.${CLASSES.EMAIL}`)).toHaveLength(0);
  });
});
