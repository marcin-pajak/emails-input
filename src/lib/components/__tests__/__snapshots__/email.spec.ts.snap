// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email template should return sanitized template 1`] = `
"
<div class=\\"EmailsInput-email Email Email--valid\\" data-id=\\"1\\">
  <span class=\\"Email-text\\">
    &lt;b>dangerous&lt;/b>@email.com
  </span>
  <button aria-label=\\"remove email\\" class=\\"Email-remove\\" data-id=\\"1\\" type=\\"button\\">
    &times;
  </button>
</div>"
`;

exports[`email template should return template for invalid email 1`] = `
"
<div class=\\"EmailsInput-email Email Email--invalid\\" data-id=\\"1\\">
  <span class=\\"Email-text\\">
    invalid
  </span>
  <button aria-label=\\"remove email\\" class=\\"Email-remove\\" data-id=\\"1\\" type=\\"button\\">
    &times;
  </button>
</div>"
`;

exports[`email template should return template for valid email 1`] = `
"
<div class=\\"EmailsInput-email Email Email--valid\\" data-id=\\"1\\">
  <span class=\\"Email-text\\">
    valid@email.com
  </span>
  <button aria-label=\\"remove email\\" class=\\"Email-remove\\" data-id=\\"1\\" type=\\"button\\">
    &times;
  </button>
</div>"
`;
